src/utils/semaphore.ts

```diff
@@ -0,0 +1,33 @@
+export class Semaphore {
+  private queue: (() => void)[] = []
+  private currentRequests = 0
+
+  constructor(
+    private maxConcurrent: number,
+    private delayMs: number
+  ) {}
+
+  async acquire(): Promise<void> {
+    if (this.currentRequests < this.maxConcurrent) {
+      this.currentRequests++
+      return
+    }
+
+    return new Promise<void>((resolve) => {
+      this.queue.push(resolve)
+    })
+  }
+
+  release(): void {
+    this.currentRequests--
+
+    if (this.queue.length > 0) {
+      const next = this.queue.shift()
+      setTimeout(() => {
+        this.currentRequests++
+        // @ts-ignore: next is guaranteed to be defined here
+        next()
+      }, this.delayMs)
```

Comment:
A few reasons:
1. Rate Limiting: By introducing a delay, you can control the rate at which requests are processed. This can be useful if you're interacting with an external service that has rate limits.
2. Preventing Starvation: If the semaphore is used in a high-concurrency environment, the delay can help prevent starvation of queued tasks by ensuring that they are given a chance to execute in a controlled manner.
3. Asynchronous Behavior: The setTimeout function ensures that the queued function is executed asynchronously, allowing the event loop to process other tasks in the meantime. This can be beneficial in a non-blocking environment like Node.js, where you want to keep the event loop free to handle other I/O operations.

Reply:
Rate limiting should not be part of "semaphore", it is just a simple counter. If you want to have it asynchronous then you can use `queueMicrotask`.